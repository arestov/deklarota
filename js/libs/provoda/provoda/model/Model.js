
import Model from '../../Model'
export default Model
