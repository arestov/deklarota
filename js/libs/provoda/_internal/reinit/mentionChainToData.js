const mentionChainLinkToData = (item) => {
  const { rel } = item
  return rel
}

export const mentionChainToData = (value) => {
  return {
    target_type: value.target_type,
    rel_path: value.list.map(mentionChainLinkToData),
    target_matcher: value.target_matcher._provoda_id,
    addr: value.addr,
    target_name: value.target_name,
    handler_payload: value.handler_payload,
  }
}
