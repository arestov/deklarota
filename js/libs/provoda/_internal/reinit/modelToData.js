const getId = (item) => item?._node_id

const modelAsValue = (someval) => {
  if (getId(someval)) {
    return { _node_id: someval._node_id }
  }

  return someval
}

export const attrValueToData = (value) => {
  if (!Array.isArray(value)) {
    return modelAsValue(value)
  }

  if (!value.some(getId)) {
    return value
  }

  return value.map(modelAsValue)
}

export const modelAttrsToData = (self) => {
  const attrs = {}

  for (let i = 0; i < self.__defined_attrs_bool.length; i++) {
    const cur = self.__defined_attrs_bool[i]
    attrs[cur.name] = attrValueToData(self.getAttr(cur.name))
  }

  return attrs
}

export const modelRelToData = (cur) => {
  if (!Array.isArray(cur)) {
    return getId(cur)
  }

  return cur.map(getId)
}

export const modelRelsToData = (self) => {
  const rels = {}

  for (const rel_name in self._extendable_nest_index) {
    if (!Object.hasOwnProperty.call(self._extendable_nest_index, rel_name)) {
      continue
    }

    rels[rel_name] = modelRelToData(self.children_models[rel_name])
  }

  return rels
}

export const mentionValueToData = (cur) => {
  const result = []

  for (const md of cur) {
    result.push(getId(md))
  }

  return result
}

export const modelMentionsToData = (self) => {
  const mentions = {}

  for (const rel_name in self.__mentions_as_rel) {
    if (!Object.hasOwnProperty.call(self.__mentions_as_rel, rel_name)) {
      continue
    }
    const cur = self.__mentions_as_rel[rel_name]
    if (cur == null) {
      continue
    }

    mentions[rel_name] = mentionValueToData(cur)
  }

  return mentions
}

const modelToData = (self) => {
  const attrs = modelAttrsToData(self)
  const rels = modelRelsToData(self)
  const mentions = modelMentionsToData(self)


  return {
    id: self._node_id,
    model_name: self.model_name,
    attrs,
    rels,
    mentions,
  }
}

export default modelToData
