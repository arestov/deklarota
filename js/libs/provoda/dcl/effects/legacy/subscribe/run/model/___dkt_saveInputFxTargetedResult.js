import saveResult from '../../../../../passes/targetedResult/save'

export default function ___dkt_saveInputFxTargetedResult(dcl, data) {
  const em = this

  saveResult(em, dcl, data, data)
}

