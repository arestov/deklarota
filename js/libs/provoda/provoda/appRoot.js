
var spv = require('spv')
var prepare = require('../structure/prepare')
var AppModel = require('./AppModel')

export default function(props, init) {
  if (typeof props == 'function') {
    if (init) {
      throw new Error('you cant pass init with Constr')
    }
    return prepare(props)
  }

  var all = {}
  if (init) {
    all.init = init
  }
  all.skip_code_path = true

  var App = spv.inh(AppModel, all, props)
  return prepare(App)

};
