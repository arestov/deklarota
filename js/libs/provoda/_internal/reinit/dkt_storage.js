import { mentionChainToData } from './mentionChainToData'
import { attrValueToData, mentionValueToData, modelAttrsToData, modelMentionsToData, modelRelsToData, modelRelToData } from './modelToData'

const noStorage = (self) => {
  return self._highway.dkt_storage == null
}

const CH_GR_LE = 2

export const createModelInDktStorage = (self) => {
  if (noStorage(self)) {
    return null
  }
  const attrs = modelAttrsToData(self)
  const rels = modelRelsToData(self)
  const mentions = modelMentionsToData(self)

  self._highway.dkt_storage.createModel(self._provoda_id, self.model_name, attrs, rels, mentions)
}

export const deleteModelInDktStorage = (self) => {
  if (noStorage(self)) {
    return null
  }
  self._highway.dkt_storage.deleteModel(self._provoda_id)

}

export const updateModelAttrsInDktStorage = (self, changes_list) => {
  if (noStorage(self)) {
    return null
  }

  const converted = new Array(changes_list.length)

  for (let i = 0; i < changes_list.length; i += CH_GR_LE) {
    const attr_name = changes_list[i]
    const value = changes_list[i + 1]

    converted[i] = attr_name
    converted[i + 1] = attrValueToData(value)
  }


  self._highway.dkt_storage.updateModelAttrs(self._provoda_id, converted)
}
export const updateModelRelInDktStorage = (self, rel_name, value) => {
  if (noStorage(self)) {
    return null
  }

  self._highway.dkt_storage.updateModelRel(self._provoda_id, rel_name, modelRelToData(value))
}

export const updateModelMentionInDktStorage = (self, mention_name, value) => {
  if (noStorage(self)) {
    return null
  }

  self._highway.dkt_storage.updateModelMention(self._provoda_id, self.model_name, mention_name, mentionValueToData(value))
}

export const createExpectedRelInDktStorage = (self, key, mention_chain) => {
  if (noStorage(self)) {
    return null
  }

  self._highway.dkt_storage.createExpectedRel(key, mentionChainToData(mention_chain))
}

export const deleteExpectedRelInDktStorage = (self, key) => {
  if (noStorage(self)) {
    return null
  }

  self._highway.dkt_storage.deleteExpectedRel(key)
}

export const commitChangesInDktStorage = (_highway) => {
  if (_highway.dkt_storage == null) {
    return null
  }

  _highway.dkt_storage.commitChanges()
}
