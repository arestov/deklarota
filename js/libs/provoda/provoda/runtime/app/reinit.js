import initApis from '../../../dcl/effects/legacy/api/init'
import memorize from '../../../../spv/memorize'
import MentionChain from '../../../Model/mentions/MentionChain'
import { setChain } from '../../../../../models/handleCurrentExpectedRel'
import { addHeavyRelQueryToRuntime } from '../../../Model/mentions/heavy_queries/addHeavyRelQuery'
import handleHeavyRelQueryChange from '../../../Model/mentions/heavy_queries/handleHeavyRelQueryChange'
import modelToData from '../../../_internal/reinit/modelToData'
import { hasOwnProperty } from '../../../hasOwnProperty'
import { countKeys } from '../../../../spv'
import updateProxy, { calcProvidedCompList } from '../../../updateProxy'
import { APP_ROOT_ID } from '../../../Model/APP_ROOT_ID'


export const reinitOne = (md) => {
  initApis(md, null)
}

const mentionChainLinkToData = (item) => {
  const { rel } = item
  return rel
}

const mentionChainToData = (value) => {
  return {
    target_type: value.target_type,
    rel_path: value.list.map(mentionChainLinkToData),
    target_matcher: value.target_matcher._provoda_id,
    addr: value.addr,
    target_name: value.target_name,
    handler_payload: value.handler_payload,
  }
}

const dataToMentionChain = (data, current_md) => {
  return new MentionChain(
    data.target_type,
    data.rel_path,
    current_md,
    null,
    null,
    data.handler_payload,
  )
}

export const toReinitableData = (runtime) => {
  const expected_rels_to_chains = runtime.expected_rels_to_chains
    ? Object.fromEntries([...runtime.expected_rels_to_chains.entries()].map(([key, value]) => {
      return [key, mentionChainToData(value)]
    }))
    : {}

  const models = {}
  for (const id in runtime.models) {
    if (!Object.hasOwnProperty.call(runtime.models, id)) {
      continue
    }
    const cur = runtime.models[id]
    if (cur == null) {
      models[id] = null
      continue
    }
    models[id] = modelToData(cur)
  }

  return {
    models,
    expected_rels_to_chains,
  }
}

const isInputRel = (rel_type) => {
  switch (rel_type) {
    case 'input':
    case 'model':
    case 'nest':
      return true
  }
  return false
}

export const getModelDataSchema = (AppRoot) => {
  const models = {}

  for (const model of AppRoot.prototype.constrs_by_name.values()) {
    const attrs = []

    for (let i = 0; i < model.__defined_attrs_bool.length; i++) {
      const cur = model.__defined_attrs_bool[i]
      const is_input = !hasOwnProperty(model.compx_check, cur.name)

      attrs.push({
        name: cur.name,
        is_input,
        is_bool: cur.type == 'bool'
      })
    }

    const rels = []
    const rels_model_schema = model._extendable_nest_index
    for (const rel_name in rels_model_schema) {
      if (!Object.hasOwnProperty.call(rels_model_schema, rel_name)) {
        continue
      }
      const cur = rels_model_schema[rel_name]
      rels.push({
        name: rel_name,
        is_input: isInputRel(cur.type),
        many: Boolean(cur.dcl.rel_shape.many)
      })
    }

    models[model.model_name] = {
      attrs,
      rels,
    }
  }

  return models
}

const getCompAttrsFromSchema = (schema) => {
  const result = new Set()

  for (let i = 0; i < schema.attrs.length; i++) {
    const cur = schema.attrs[i]
    if (cur.is_input) {
      continue
    }
    result.add(cur.name)
  }

  return result
}

const makeAutomigration = (old_schema, new_schema) => {
  /*
    now we can only auto migrate to added comp attrs

    p.s.
    1. it would be easy to make
      - remove comp attrs
      - add/remove comp rels

    2. medium - signal to recalc exising comp attrs

    3. rename comp (just remove + add)

    4. most hard - rename input attrs/rels?
  */
  const result = {}

  for (const model_name in new_schema) {
    if (!Object.hasOwnProperty.call(new_schema, model_name)) {
      continue
    }

    /*
      there is no model in old schema. new model added. nothing to compare
    */
    if (!hasOwnProperty(old_schema, model_name)) {
      continue
    }


    const old_schema_comp_attrs = getCompAttrsFromSchema(old_schema[model_name])
    const new_schema_comp_attrs = getCompAttrsFromSchema(new_schema[model_name])

    const added = []
    for (const attr_name of new_schema_comp_attrs) {
      if (old_schema_comp_attrs.has(attr_name)) {
        continue
      }
      added.push(attr_name)
    }
    if (!added.length) {
      continue
    }

    result[model_name] = {
      added_comp_attrs: added,
    }
  }

  if (!countKeys(result, true)) {
    return null
  }

  return result
}

const reinitAllAttrsAutoMigration = (new_schema) => {
  const result = {}

  for (const model_name in new_schema) {
    if (!Object.hasOwnProperty.call(new_schema, model_name)) {
      continue
    }

    const new_schema_comp_attrs = getCompAttrsFromSchema(new_schema[model_name])
    result[model_name] = {
      added_comp_attrs: [...new_schema_comp_attrs],
    }
  }


  return result

}

export const reinit = async (AppRoot, runtime, data, interfaces, options) => {
  /* expect reinit_all_attrs to be used in dev mode, not production */
  const reinit_all_attrs = options?.reinit_all_attrs
  const {models, expected_rels_to_chains} = data

  const models_list = []
  for (const id in models) {
    if (!Object.hasOwnProperty.call(models, id)) {
      continue
    }

    const cur = models[id]
    if (cur == null) {
      runtime.models[id] = null
      continue
    }

    runtime.models_counters = Math.max(runtime.models_counters, cur.id + 1)

    const Constr = AppRoot.prototype.constrs_by_name.get(cur.model_name).constructor
    const opts = {
      _highway: runtime,
      _provoda_id: cur.id,
      map_parent: null,
      app: null,
      reinit: true,
    }

    const item = new Constr(opts)
    if (runtime.models[item._provoda_id] !== item) {
      throw new Error()
    }

    models_list.push(cur)
  }

  const getById = (id) => runtime.models[id]

  const listToRefs = (list) => list.map(getById)

  const toRefs = (value) => {
    if (value == null) {
      return value
    }

    if (!Array.isArray(value)) {
      return getById(value)
    }

    return listToRefs(value)
  }

  const hasProvodaId = (item) => Boolean(item?._provoda_id)

  const getByProvodaId = (item) => getById(item._provoda_id)

  const toRuntimeAttrValue = (val) => {
    if (!Array.isArray(val)) {
      if (!hasProvodaId(val)) {
        return val
      }

      return getById(val._provoda_id)
    }
    if (!val.some(hasProvodaId)) {
      return val
    }
    return val.map(getByProvodaId)
  }

  const app = getById(APP_ROOT_ID)


  for (let i = 0; i < models_list.length; i++) {
    const cur = models_list[i]
    const self = runtime.models[cur.id]

    for (const rel_name in cur.rels) {
      if (!Object.hasOwnProperty.call(cur.rels, rel_name)) {
        continue
      }

      self.children_models[rel_name] = toRefs(cur.rels[rel_name])
    }

    self.map_parent = self.children_models.$parent || null
    self.app = app

    self.__mentions_as_rel = {}
    for (const rel_name in cur.mentions) {
      if (!Object.hasOwnProperty.call(cur.mentions, rel_name)) {
        continue
      }

      self.__mentions_as_rel[rel_name] = new Set(listToRefs(cur.mentions[rel_name]))
    }

    for (const attr_name in cur.attrs) {
      if (!Object.hasOwnProperty.call(cur.attrs, attr_name)) {
        continue
      }

      const value = cur.attrs[attr_name]
      self.states[attr_name] = toRuntimeAttrValue(value)
    }
  }

  const expected_rels = []
  for (const key in expected_rels_to_chains) {
    if (!Object.hasOwnProperty.call(expected_rels_to_chains, key)) {
      continue
    }

    const element = expected_rels_to_chains[key]
    const curremt_md_id = element.target_matcher
    const current_md = getById(curremt_md_id)
    const expected_rel = element.handler_payload.data

    const chain = dataToMentionChain(element, current_md)

    setChain(current_md, expected_rel, chain)
    addHeavyRelQueryToRuntime(current_md, chain)

    expected_rels.push(
      [current_md, chain]
    )
  }

  const old_schema = await runtime.dkt_storage?.getSchema()
  const new_schema = getModelDataSchema(AppRoot)

  const buildAutoMigration = () => {
    if (reinit_all_attrs) {
      return reinitAllAttrsAutoMigration(new_schema)
    }
    if (old_schema) {
      return makeAutomigration(old_schema, new_schema)
    }

    return null
  }

  const auto_migration = buildAutoMigration()

  if (runtime.dkt_storage != null) {
    runtime.dkt_storage.putSchema(new_schema)
  }

  runtime.calls_flow.input(() => {
    const isApiAttr = memorize((str) => str.startsWith('$meta$apis$') && str.endsWith('$used'))

    const initModelAttrs = (self, cur, auto_migration) => {
      /*
        1. add/remove new attrs in schema
        2. acknowledge apis where removed
      */

      const changes = []

      /*
      1.
      */
      const added_comp_attrs = auto_migration?.[self.model_name]?.added_comp_attrs
      if (added_comp_attrs) {
        calcProvidedCompList(self, added_comp_attrs, changes)
      }

      /*
      2
      */
      for (const attr_name in cur.attrs) {
        if (!Object.hasOwnProperty.call(cur.attrs, attr_name)) {
          continue
        }

        if (!isApiAttr(attr_name)) {
          continue
        }

        changes.push(attr_name, false)
      }

      if (!changes.length) {
        return
      }

      updateProxy(self, changes)
    }

    for (let i = 0; i < models_list.length; i++) {
      const cur = models_list[i]
      const self = runtime.models[cur.id]
      initModelAttrs(self, cur, auto_migration)
    }

  })

  runtime.calls_flow.input(() => {

    for (let i = 0; i < models_list.length; i++) {
      const cur = models_list[i]
      const self = runtime.models[cur.id]
      reinitOne(self)
    }

    const app = getById(APP_ROOT_ID)

    for (const api_name in interfaces) {
      if (!Object.hasOwnProperty.call(interfaces, api_name)) {
        continue
      }

      const cur = interfaces[api_name]
      app.useInterface(api_name, cur)
    }
  })

  runtime.calls_flow.input(() => {
    expected_rels.forEach(([current_md, chain]) => {
      handleHeavyRelQueryChange(current_md, chain)
    })
  })

  return new Promise(resolve => {
    runtime.calls_flow.input(() => {

      const app_model = getById(APP_ROOT_ID)

      resolve({
        runtime,
        app_model: app_model,
        flow: runtime.calls_flow,
        sync_sender: runtime.sync_sender,
        views_proxies: runtime.views_proxies,
      })
    })

  })

}
