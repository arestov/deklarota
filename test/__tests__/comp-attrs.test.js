/* eslint-disable fp/no-let */
/* eslint-disable fp/no-loops */
/* eslint-disable fp/no-delete */
/* eslint-disable fp/no-mutation */

import { it } from '@jest/globals'
import appRoot from 'pv/appRoot.js'
import model from 'pv/model'
import mergeModel from 'pv/dcl/merge'
import testingInit, { testingReinit } from '../testingInit'
import { toReinitableData } from '../../js/libs/provoda/provoda/runtime/app/reinit'

it('should compute comp attr before & after reinit', async () => {
  const makeStorage = () => {
    const storage_data = {
      models: {},
      expected_rels_to_chains: {},
    }

    let schema = null

    const storage = {
      getSchema: () => schema,
      putSchema: val => {
        schema = val
      },

      hasData: () => Boolean(Object.keys(storage_data.models).length),
      getSnapshot: () => storage_data,
      createModel: (id, model_name, attrs, rels, mentions) => {
        storage_data.models[id] = {
          id, model_name, attrs, rels, mentions,
        }
      },
      deleteModel: id => {
        storage_data.models[id] = null
      },
      updateModelAttrs: (id, changes_list) => {
        const CH_GR_LE = 2
        for (let i = 0; i < changes_list.length; i += CH_GR_LE) {
          const attr_name = changes_list[i]
          const value = changes_list[i + 1]
          storage_data.models[id].attrs[attr_name] = value
        }
      },
      updateModelRel: (id, rel_name, value) => {
        storage_data.models[id].rels[rel_name] = value
      },
      updateModelMention: (id, mention_name, value) => {
        storage_data.models[id].mentions[mention_name] = value
      },
      createExpectedRel: (key, data) => {
        storage.expected_rels_to_chains[key] = data
      },
      deleteExpectedRel: key => {
        delete storage.expected_rels_to_chains[key]
      },
    }
    return storage
  }

  const dkt_storage = makeStorage()


  const Track = model({
    model_name: 'track',
    attrs: {
      number: ['input'],
      allowedToPlayType: [
        'comp',
        ['< @one:playlistModeB < $parent', '< @one:rootModeB < $root', 'number', '< @one:$meta$rels$tracks$length < $parent'],
        (parentMark, rootMark, number, length) => {
          if (number > (2 / 3 * length)) {
            return 'independent'
          }

          if (number > (1 / 3 * length)) {
            if (!rootMark) {
              return 'rootA'
            }
            return 'rootB'
          }

          if (!parentMark) {
            return 'parentA'
          }

          return 'parentB'
        },
      ],
    },
  })

  const Playlist = model({
    model_name: 'Playlist',
    attrs: {
      playlistModeB: ['input'],
    },
    rels: {
      tracks: ['model', Track, { many: true }],
    },
    actions: {
      initTracks: {
        to: ['<< tracks', { method: 'at_end', can_create: true }],
        fn: list => list.map(attrs => ({ attrs })),
      },
    },
  })

  const AppRootSchema = {
    attrs: {
      rootModeB: ['input'],
      resultAttr: ['comp', ['< @all:allowedToPlayType < playlist.tracks']],
    },
    rels: {
      playlist: ['nest', [Playlist]],
    },
  }
  const AppRoot = appRoot(AppRootSchema)

  const inited = await testingInit(AppRoot, {}, { dkt_storage })
  expect(dkt_storage.getSchema()).toMatchSnapshot()

  {
    inited.app_model.getNesting('playlist').dispatch('initTracks', [
      { number: 1 },
      { number: 2 },
      { number: 3 },
    ])

    await inited.computed()

    expect(inited.app_model.getAttr('resultAttr')).toMatchSnapshot()

    const data = toReinitableData(inited.app_model._highway)
    expect(dkt_storage.getSnapshot()).toEqual(data)

    {
      const reinited = await testingReinit(AppRoot, data, {})
      const inited = reinited
      expect(inited.app_model.getAttr('resultAttr')).toMatchSnapshot()


      /*
        2. change state
      */
      inited.app_model.updateAttr('rootModeB', true)
      inited.app_model.getNesting('playlist').updateAttr('playlistModeB', true)


      /*
        3. expect all changed propagated
      */
      await inited.computed()
      expect(inited.app_model.getAttr('resultAttr')).toMatchSnapshot()
    }

    {
      const AppRootChanged = appRoot(
        mergeModel(AppRootSchema, {
          attrs: {
            oneMoreAttr: ['comp', ['resultAttr']],
          },
        }),
      )

      const reinited = await testingReinit(AppRootChanged, data, {}, { dkt_storage })
      /* compare with resultAttr */
      expect(
        reinited.app_model.getAttr('oneMoreAttr'),
      ).toBe(
        reinited.app_model.getAttr('resultAttr'),
      )

      /* compare with original rootModeB */
      expect(
        reinited.app_model.getAttr('oneMoreAttr'),
      ).toBe(
        inited.app_model.getAttr('resultAttr'),
      )
    }
  }
})
