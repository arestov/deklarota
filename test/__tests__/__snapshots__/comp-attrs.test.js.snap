// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should compute comp attr before & after reinit 1`] = `
Object {
  "Playlist": Object {
    "attrs": Array [
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$inited",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$removed",
      },
      Object {
        "is_bool": true,
        "is_input": false,
        "name": "$meta$fxs_out$__remove_model$deps_ready",
      },
      Object {
        "is_bool": true,
        "is_input": false,
        "name": "__fx_out_triggering_attrs",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$$root$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$$root$exists",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$$parent$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$$parent$exists",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$tracks$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$tracks$exists",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "_node_id",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$inited",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$removed",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "<<<<",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$apis$self$used",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "url_part",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "nav_title",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$perspective$each_show",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$perspective$each_focus",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "mp_has_focus",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "mp_show",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "playlistModeB",
      },
    ],
    "rels": Array [
      Object {
        "is_input": true,
        "many": false,
        "name": "$root",
      },
      Object {
        "is_input": true,
        "many": false,
        "name": "$parent",
      },
      Object {
        "is_input": true,
        "many": true,
        "name": "tracks",
      },
    ],
  },
  "app_model": Object {
    "attrs": Array [
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$inited",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$removed",
      },
      Object {
        "is_bool": false,
        "is_input": false,
        "name": "resultAttr",
      },
      Object {
        "is_bool": true,
        "is_input": false,
        "name": "$meta$fxs_out$__remove_model$deps_ready",
      },
      Object {
        "is_bool": true,
        "is_input": false,
        "name": "__fx_out_triggering_attrs",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$$root$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$$root$exists",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$$parent$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$$parent$exists",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$$session_root$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$$session_root$exists",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$playlist$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$playlist$exists",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$playlist$autoinited",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "_node_id",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$inited",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$removed",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "<<<<",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$apis$self$used",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "< @all:allowedToPlayType < playlist.tracks <<",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "url_part",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "nav_title",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$perspective$each_show",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$perspective$each_focus",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "mp_has_focus",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "mp_show",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "rootModeB",
      },
    ],
    "rels": Array [
      Object {
        "is_input": true,
        "many": false,
        "name": "$root",
      },
      Object {
        "is_input": true,
        "many": false,
        "name": "$parent",
      },
      Object {
        "is_input": true,
        "many": false,
        "name": "$session_root",
      },
      Object {
        "is_input": true,
        "many": false,
        "name": "playlist",
      },
    ],
  },
  "root_bwlev": Object {
    "attrs": Array [
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$inited",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$removed",
      },
      Object {
        "is_bool": true,
        "is_input": false,
        "name": "$meta$fxs_out$__remove_model$deps_ready",
      },
      Object {
        "is_bool": true,
        "is_input": false,
        "name": "__fx_out_triggering_attrs",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$$root$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$$root$exists",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$$parent$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$$parent$exists",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$pioneer$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$pioneer$exists",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$spyglasses$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$spyglasses$exists",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "_node_id",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$inited",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$removed",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "<<<<",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$apis$self$used",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "used_data_structure",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "spyglasses_index",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "spyglasses_requests",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "probe_name",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "map_level_num",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "is_main_perspectivator_resident",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "pioneer_node_id",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "pioneer",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "freeze_parent_bwlev",
      },
    ],
    "rels": Array [
      Object {
        "is_input": true,
        "many": false,
        "name": "$root",
      },
      Object {
        "is_input": true,
        "many": false,
        "name": "$parent",
      },
      Object {
        "is_input": true,
        "many": false,
        "name": "pioneer",
      },
      Object {
        "is_input": true,
        "many": true,
        "name": "spyglasses",
      },
    ],
  },
  "track": Object {
    "attrs": Array [
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$inited",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$removed",
      },
      Object {
        "is_bool": false,
        "is_input": false,
        "name": "< @one:playlistModeB < $parent <<",
      },
      Object {
        "is_bool": false,
        "is_input": false,
        "name": "< @one:rootModeB < $root <<",
      },
      Object {
        "is_bool": false,
        "is_input": false,
        "name": "< @one:$meta$rels$tracks$length < $parent <<",
      },
      Object {
        "is_bool": false,
        "is_input": false,
        "name": "allowedToPlayType",
      },
      Object {
        "is_bool": true,
        "is_input": false,
        "name": "$meta$fxs_out$__remove_model$deps_ready",
      },
      Object {
        "is_bool": true,
        "is_input": false,
        "name": "__fx_out_triggering_attrs",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$$root$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$$root$exists",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$rels$$parent$length",
      },
      Object {
        "is_bool": true,
        "is_input": true,
        "name": "$meta$rels$$parent$exists",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "_node_id",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$inited",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$removed",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "<<<<",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$apis$self$used",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "< @all:playlistModeB < $parent <<",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "< @all:rootModeB < $root <<",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "< @all:$meta$rels$tracks$length < $parent <<",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "url_part",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "nav_title",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$perspective$each_show",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "$meta$perspective$each_focus",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "mp_has_focus",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "mp_show",
      },
      Object {
        "is_bool": false,
        "is_input": true,
        "name": "number",
      },
    ],
    "rels": Array [
      Object {
        "is_input": true,
        "many": false,
        "name": "$root",
      },
      Object {
        "is_input": true,
        "many": false,
        "name": "$parent",
      },
    ],
  },
}
`;

exports[`should compute comp attr before & after reinit 2`] = `
Array [
  "parentA",
  "rootA",
  "independent",
]
`;

exports[`should compute comp attr before & after reinit 3`] = `
Array [
  "parentA",
  "rootA",
  "independent",
]
`;

exports[`should compute comp attr before & after reinit 4`] = `
Array [
  "parentB",
  "rootB",
  "independent",
]
`;
